File: Chalkable.Web/app/chlk/models/announcement/ClassAnnouncementViewData.js
REQUIRE('ria.serialize.SJX');
REQUIRE('chlk.models.announcement.AnnouncementWithExpiresDateViewData');
REQUIRE('chlk.models.common.ChlkDate');

REQUIRE('chlk.models.id.ClassId');

NAMESPACE('chlk.models.announcement', function () {
    "use strict";

    var SJX = ria.serialize.SJX;

    /** @class chlk.models.announcement.ClassAnnouncementViewData*/
    CLASS(
        UNSAFE, 'ClassAnnouncementViewData',
                EXTENDS(chlk.models.announcement.AnnouncementWithExpiresDateViewData),
                IMPLEMENTS(ria.serialize.IDeserializable), [

            OVERRIDE, VOID, function deserialize(raw) {
                BASE(raw);
                this.defaultTitle = SJX.fromValue(raw.defaulttitle, Boolean);
                this.setAnnouncementTypeId(SJX.fromValue(raw.announcementtypeid, Number));
                this.classId = SJX.fromValue(Number(raw.classid), chlk.models.id.ClassId);
                this.shortClassName = SJX.fromValue(raw.classname, String);
                this.className = SJX.fromValue(raw.fullclassname, String);
                this.dropped = SJX.fromValue(raw.dropped, Boolean);
                this.order = SJX.fromValue(raw.order, Number);
                this.maxScore = SJX.fromValue(raw.maxscore, Number);
                this.ableDropStudentScore = SJX.fromValue(raw.candropstudentscore, Boolean);
                this.ableToExempt = SJX.fromValue(raw.maybeexempt, Boolean);
                this.gradable = SJX.fromValue(raw.gradable, Boolean);
                this.ableToGrade = SJX.fromValue(raw.cangrade, Boolean);
                this.hiddenFromStudents = SJX.fromValue(raw.hidefromstudents, Boolean);
                this.weightMultiplier = SJX.fromValue(raw.weightmultiplier, Number);
                this.weightAddition = SJX.fromValue(raw.weightaddition, Number);
            },

            function $(){
                BASE();
                this.announcementTypeId = null;
                this.classId = null;
                this._annTypeEnum = chlk.models.announcement.AnnouncementTypeEnum;
            },

            String, 'defaultTitle',
            Number, 'announcementTypeId',
            chlk.models.id.ClassId, 'classId',
            String, 'shortClassName',
            String, 'className',
            Boolean, 'dropped',
            Number, 'order',
            Number, 'maxScore',
            Boolean, 'ableDropStudentScore',
            Boolean, 'ableToExempt',
            Boolean, 'gradable',
            Boolean, 'ableToGrade',
            Boolean, 'hiddenFromStudents',
            Number, 'weightMultiplier',
            Number, 'weightAddition',

            Number, 'avg'
        ]);
});

File: Chalkable.Web/Models/AnnouncementsViewData/ClassAnnouncementViewData.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Chalkable.BusinessLogic.Model;
using Chalkable.Data.School.Model.Announcements;

namespace Chalkable.Web.Models.AnnouncementsViewData
{
    public class ClassAnnouncementViewData : ShortAnnouncementViewData
    {
        public DateTime? ExpiresDate { get; set; }
        public string DefaultTitle { get; set; }
        public int? AnnouncementTypeId { get; set; }
        public int? ChalkableAnnouncementTypeId { get; set; }
        public int? ClassId { get; set; }
        public string ClassName { get; set; }
        public string FullClassName { get; set; }
        public Guid? DepartmentId { get; set; }
        public bool Dropped { get; set; }
        public int Order { get; set; }
        public decimal? MaxScore { get; set; }
        public bool CanDropStudentScore { get; set; }
        public bool MayBeExempt { get; set; }
        public bool Gradable { get; set; }
        public bool CanGrade { get; set; }
        public bool HideFromStudents { get; set; }

        public decimal? WeightMultiplier { get; set; }
        public decimal? WeightAddition { get; set; }

        protected ClassAnnouncementViewData(ClassAnnouncement announcement, IList<ClaimInfo> claims)
            : base(announcement)
        {
            DefaultTitle = announcement.DefaultTitle;
            AnnouncementTypeId = announcement.ClassAnnouncementTypeRef;
            ChalkableAnnouncementTypeId = announcement.ChalkableAnnouncementType;
            PersonId = announcement.PrimaryTeacherRef;
            PersonName = announcement.PrimaryTeacherName;
            PersonGender = announcement.PrimaryTeacherGender;
            ClassId = announcement.ClassRef;
            ClassName = announcement.ClassName;
            FullClassName = announcement.FullClassName;
            Dropped = announcement.Dropped;
            ExpiresDate = announcement.Expires == DateTime.MinValue ? (DateTime?)null : announcement.Expires;
            Order = announcement.Order;
            IsOwner = announcement.IsOwner;
            MaxScore = announcement.MaxScore;
            CanDropStudentScore = announcement.MayBeDropped;
            MayBeExempt = announcement.MayBeExempt;
            Gradable = announcement.IsScored;
            CanGrade = (IsOwner || claims.HasPermission(ClaimInfo.MAINTAIN_CLASSROOM_ADMIN)) && announcement.IsScored;
            DepartmentId = announcement.DepartmentId;
            HideFromStudents = !announcement.VisibleForStudent;
            WeightAddition = announcement.WeightAddition;
            WeightMultiplier = announcement.WeightMultiplier;
        }

        public static ClassAnnouncementViewData Create(ClassAnnouncement announcement, IList<ClaimInfo> claims)
        {
            return new ClassAnnouncementViewData(announcement, claims);
        }
        public static IList<ClassAnnouncementViewData> Create(IList<ClassAnnouncement> announcements, IList<ClaimInfo> claims)
        {
            return announcements.Select(x => Create(x, claims)).ToList();
        }
    }

}